### PR TITLE
ARROW-6760: [C++] More informative error messages for JSON parsing errors

### PR DESCRIPTION
**Pull request by bkietz**
Parse errors will now include the path to whichever field caused the parse error and the row number where the parse error occurred.

### CODE REVIEW

**Comment by bkietz** [comment]
@ursabot build

File: cpp/src/arrow/json/parser_test.cc [hunk 1]
```diff
@@ -144,14 +144,48 @@ TEST(BlockParserWithSchema, SkipFieldsOutsideSchema) {
...
+}
+
+TEST_P(BlockParserTypeError, FailOnNestedInconvertible) {
+  auto options = Options(schema({field("a", list(struct_({field("a", int32())})))}));
+  std::shared_ptr<Array> parsed;
+  Status error =
+      ParseFromString(options, "{\"a\":[{\"a\":0}]}\n{\"a\":[{\"a\":true}]}", &parsed);
+  ASSERT_RAISES(Invalid, error);
+  ASSERT_EQ(error.message(),
+            "JSON parse error: Column(/a/[]/a) changed from number to boolean in row 1");
```
Comment (pitrou) [inline cpp/src/arrow/json/parser_test.cc]: Test with two different field names perhaps? (so that we know that we get e.g. "/a/[]/b" and not "/b/[]/a")

**Review by pitrou — COMMENTED** [review]
LGTM, just a nit

File: cpp/src/arrow/json/parser_test.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/json/parser_test.cc]: will do

File: cpp/src/arrow/json/parser_test.cc [hunk 2]
```diff
@@ -144,14 +145,54 @@ TEST(BlockParserWithSchema, SkipFieldsOutsideSchema) {
...
+  ASSERT_RAISES(Invalid, error);
+  EXPECT_THAT(
+      error.message(),
+      testing::StartsWith("JSON parse error: Column(/a) was specified twice in row 0"));
 }
 
+INSTANTIATE_TEST_CASE_P(BlockParserTypeError, BlockParserTypeError,
+                        ::testing::Values(UnexpectedFieldBehavior::Ignore,
+                                          UnexpectedFieldBehavior::Ignore,
+                                          UnexpectedFieldBehavior::InferType));
```
Comment (pitrou) [inline cpp/src/arrow/json/parser_test.cc]: Why is "Ignore" given twice here?

**Review by pitrou — APPROVED** [review]
Looks great. One last question.

File: cpp/src/arrow/json/parser_test.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/json/parser_test.cc]: Typo. One of those should be Error

**Pull request closed without merging**